### PR TITLE
Exponential domain

### PR DESCRIPTION
**Pull request by bksaiki**
This PR adds the exponential (`exp`, `log`, `pow`, etc.) domain. It is bootstrapped from the rational domain using a set of axioms about the exponential functions.

### CODE REVIEW

File: tests/exponential.rs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        F: FnMut(&'a Id) -> &'a CVec<Self>,
+    {
+        vec![]
+    }
+
+    fn initialize_vars(egraph: &mut EGraph<Self, SynthAnalysis>, vars: &[String]) {
+        for v in vars {
+            let id = egraph.add(Exponential::Var(Symbol::from(v.clone())));
+
+            let l_id = egraph.add(Exponential::Log(id));
```
Comment (ajpal) [inline tests/exponential.rs:48]: Why are you adding these here? shouldn't these be in the workload?

Comment (bksaiki) [reply tests/exponential.rs:48]: Oh I forgot to union those expressions! Is there a better way to assert that `?x <==> (exp (log ?x))`  (and `?x <==> (log (exp ?x))` only for variables?

**Comment by bksaiki** [comment]
Debugging from Herbie:
```
Registering ruleset `exponential.json` ...
   Groups: (arithmetic)
   Vars:   ((c . real) (a . real) (b . real))
   Rules:  2
    exponential.json-1: 1 => (sqrt 1)
    exponential.json-2: 1 => (cbrt 1)
  Registering ruleset `exponential.json-simplify` ...
   Groups: (simplify arithmetic)
   Vars:   ((c . real) (a . real) (b . real))
   Rules:  29
    exponential.json-simplify-1: (exp (+ a b)) => (* (exp a) (exp b))
    exponential.json-simplify-2: (exp (neg a)) => (/ 1 (exp a))
    exponential.json-simplify-3: (* (exp a) (exp b)) => (exp (+ a b))
    exponential.json-simplify-4: (/ 1 (exp a)) => (exp (neg a))
    exponential.json-simplify-5: (exp 0) => 1
    exponential.json-simplify-6: (log (exp a)) => a
    exponential.json-simplify-7: (exp (log a)) => a
    exponential.json-simplify-8: (pow 1 a) => 1
    exponential.json-simplify-9: (pow a 1) => a
    exponential.json-simplify-10: (* 1/2 (log a)) => (log (sqrt a))
    exponential.json-simplify-11: (log (cbrt a)) => (* 1/3 (log a))
    exponential.json-simplify-12: (sqrt (cbrt a)) => (cbrt (sqrt a))
    exponential.json-simplify-13: (cbrt (* b a)) => (* (cbrt b) (cbrt a))
    exponential.json-simplify-14: (* (sqrt b) (sqrt a)) => (sqrt (* a b))
    exponential.json-simplify-15: (pow (* b b) (log a)) => (pow (* a a) (log b))
    exponential.json-simplify-16: (pow (cbrt b) (log a)) => (pow (cbrt a) (log b))
    exponential.json-simplify-17: (pow (exp b) a) => (exp (* b a))
    exponential.json-simplify-18: (pow b (log a)) => (pow a (log b))
    exponential.json-simplify-19: (cbrt (pow b a)) => (pow (cbrt b) a)
    exponential.json-simplify-20: (pow (sqrt b) a) => (sqrt (pow b a))
    exponential.json-simplify-21: (* (* c b) (log (* a a))) => (* (+ c c) (log (pow a b)))
    exponential.json-simplify-22: (pow c (+ b a)) => (* (pow c b) (pow c a))
    exponential.json-simplify-23: (* (pow c a) (pow b a)) => (pow (* b c) a)
    exponential.json-simplify-24: (pow (exp c) (* b a)) => (pow (exp b) (* a c))
    exponential.json-simplify-25: (* c (log (pow b a))) => (* (log b) (* a c))
    exponential.json-simplify-26: (* c (log (pow b a))) => (* a (log (pow b c)))
    exponential.json-simplify-27: (pow (pow c b) (log a)) => (pow (pow a b) (log c))
    exponential.json-simplify-28: (pow (pow c b) a) => (pow c (* b a))
    exponential.json-simplify-29: (pow (pow c b) a) => (pow (pow c a) b)
```

File: tests/exponential.rs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        ValidationResult::Valid
+    }
+
+    fn is_rule_lifting() -> bool {
+        true
+    }
+
+    fn get_lifting_rules() -> enumo::Ruleset<Self> {
+        enumo::Ruleset::from_str_vec(&[
+            // definitions (denote)
```
Comment (ajpal) [inline tests/exponential.rs:91]: I'm still concerned about these in terms of the allowed/forbidden narrative, but it's fine for now

File: tests/exponential.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+            "(exp 0) ==> 1",
+            // inverse properties
+            "(log (exp ?a)) ==> ?a",
+            "(exp (log ?a)) ==> ?a",
+        ])
+    }
+
+    // TODO: actually derive these rules
+    fn rational_rules() -> Ruleset {
+        Ruleset::from_str_vec(&[
```
Comment (ajpal) [inline tests/exponential.rs:137]: what are these? could you use baseline/rational.rules? or add a new file to baseline/ ?

**Review by ajpal — APPROVED** [review]

File: tests/exponential.rs (hunk 3 above)
Comment (bksaiki) [reply tests/exponential.rs:137]: Oh yep. Forgot about that.

**Pull request merged**